### PR TITLE
[LMLayer, Web] Context Management + Suggestion UX

### LINKED ISSUE #1849: [Web] Various tweaks to predictive text
- [ ] `<keep>` should not appear when 0 things typed (empty buffer). It should show top 3 words!
```
      <keep> => <inserts a space>
      <space> => <fat-finger suggestions>
```
- [ ] model compositor does not take out duplicates
- [ ] backspace like a normal backspace key — not the entire suggestion
- [ ] `<keep>` should disappear when suggestion is same as verbatim

### PR DESCRIPTION
**Pull request by jahorton**
Fixes #1849.

This PR performs the following enhancements:
- [x] Adds a wordbreaking API to the LMLayer for use by KMW.
- [x] LMLayer now generates 'keep' `Suggestion`s and prevents `Suggestion` duplication.
- [x] KMW now relies on the new 'keep' `Suggestion` from LMLayer; no more `<keep>` text!
- [x] The LMLayer now properly processes BKSP keystrokes.  (The fixed 'bug'.)
- [x] KMW now requests wordbreaking to get display text for reversion operations.
- [x] On BKSP, KMW displays a reversion option instead of auto-reverting.
- [x] BKSP is no longer blocked post-acceptance, while still allowing reversion.

### CODE REVIEW

**Comment by keyman-server** [comment]
[OSK renders for Windows](https://build.palaso.org/repository/download/KeymanDesktop_TestPrRenderOnScreenKeyboards/155315:id/kvk/index.html)

File: common/predictive-text/message.d.ts [hunk 1]
```diff
@@ -41,8 +41,8 @@ type Token = number;
 /**
  * The valid outgoing message kinds.
  */
-type OutgoingMessageKind = 'ready' | 'suggestions';
-type OutgoingMessage = ReadyMessage | SuggestionMessage;
+type OutgoingMessageKind = 'ready' | 'suggestions' | 'word';
+type OutgoingMessage = ReadyMessage | SuggestionMessage | BrokenWordMessage;
```
Comment (eddieantonio) [inline common/predictive-text/message.d.ts]: Since `ReadyMessage` uses `'ready` as its message type, and `SuggestionMessage` uses `*'suggestions'` as its message type, I'm wondering if we can do the same with `BrokenWordMessage` (and `SuggestionsMessage` for that matter! 😂)

For example,

```suggestion
type OutgoingMessage = ReadyMessage | SuggestionMessage | WordMessage;
```

```suggestion
type OutgoingMessage = ReadyMessage | SuggestionMessage | CurrentWordMessage;
```

...but that would incur changing the message type itself to `currentword` :/

File: common/predictive-text/message.d.ts [hunk 2]
```diff
@@ -72,6 +72,25 @@ interface SuggestionMessage {
   suggestions: Suggestion[];
 }
 
+/**
+ * Sends the keyboard an ordered list of suggestions.
+ */
+interface BrokenWordMessage {
+  message: 'word';
```
Comment (eddieantonio) [inline common/predictive-text/message.d.ts]: See comment above about interface name and tag name consistency.

File: common/predictive-text/unit_tests/headless/default-word-breaker.js [hunk 3]
```diff
@@ -19,4 +20,63 @@ describe('The default word breaker', function () {
...
+    // A limitation of the current implementation; we should fix this before release.
+    // Then again, when typing this is probably fine; just not when not typing.
+    context = { 
+      left: 'The quick brown fox jum', startOfBuffer: true,
+      right: 'ped over the lazy dog.', endOfBuffer: true
+    };
+
+    broken = model.wordbreak(context);
+
+    assert.strictEqual(broken, 'jumped');  // Current result:  'jum'.
```
Comment (eddieantonio) [inline common/predictive-text/unit_tests/headless/default-word-breaker.js]: Ah! So you're proposing "current word" semantics! That's great! I think I know how to implement that! This is a big step up from SwiftKey which strictly considers only what's left of the cursor.

File: common/predictive-text/unit_tests/headless/default-word-breaker.js [hunk 4]
```diff
@@ -19,4 +20,63 @@ describe('The default word breaker', function () {
...
+    // Wordbreaking on a empty space => no word.
+    context = { 
+      left: 'The quick brown fox jumped ', startOfBuffer: true,
+      right: 'over the lazy dog.', endOfBuffer: true
+    };
+
+    broken = model.wordbreak(context);
+
+    assert.strictEqual(broken, '');
+  });
```
Comment (eddieantonio) [inline common/predictive-text/unit_tests/headless/default-word-breaker.js]: One more corner-case test may be nice: what should be returned when the buffer is completely empty?

File: common/predictive-text/unit_tests/in_browser/cases/worker-dummy-integration.js [hunk 5]
```diff
@@ -45,6 +45,41 @@ describe('LMLayer using dummy model', function () {
     });
   });
 
+  describe('Wordbreaking', function () {
+    it('will perform (default) wordbreaking and return word at caret', function () {
+      if(navigator.userAgent.indexOf('MSIE') !== -1 || navigator.appVersion.indexOf('Trident/') > -1) {
```
Comment (eddieantonio) [inline common/predictive-text/unit_tests/in_browser/cases/worker-dummy-integration.js]: 👍 
---
🤦‍♂ 🌎

File: common/predictive-text/worker/model-compositor.ts [hunk 6]
```diff
@@ -9,12 +9,36 @@ class ModelCompositor {
   predict(transformDistribution: Transform | Distribution<Transform>, context: Context): Suggestion[] {
     let suggestionDistribution: Distribution<Suggestion> = [];
 
+    // Assumption:  Duplicated 'displayAs' properties indicate duplicated Suggestions.
+    // When true, we can use an associated array to de-duplicate everything.
+    let suggestionDistribMap: {[key: string]: ProbabilityMass<Suggestion>} = {};
```
Comment (eddieantonio) [inline common/predictive-text/worker/model-compositor.ts]: 👍

File: common/predictive-text/worker/model-compositor.ts [hunk 7]
```diff
@@ -9,12 +9,36 @@ class ModelCompositor {
...
       transformDistribution = [ {sample: transformDistribution, p: 1.0} ];
     }
 
+    // Find the transform for the actual keypress.
+    let inputTransform = transformDistribution.sort(function(a, b) {
+      return b.p - a.p;
+    })[0].sample;
+
+    // Only allow new-word suggestions if space was the most likely keypress.
+    let allowSpace = inputTransform.insert == " " || inputTransform.insert == "\n";
```
Comment (eddieantonio) [inline common/predictive-text/worker/model-compositor.ts]: Should this be extended to other, non-ASCII whitespace?

File: common/predictive-text/worker/model-compositor.ts [hunk 8]
```diff
@@ -24,15 +48,46 @@ class ModelCompositor {
           pair.sample.transformId = transform.id;
         }
 
-        let compositedPair = {sample: pair.sample, p: pair.p * alt.p};
-        suggestionDistribution.push(compositedPair);
+        // Combine duplicate samples.
+        let displayText = pair.sample.displayAs;
+
+        if(displayText == keepOptionText) {
```
Comment (eddieantonio) [inline common/predictive-text/worker/model-compositor.ts]: 👍 This is great!

File: common/predictive-text/worker/model-compositor.ts [hunk 9]
```diff
@@ -24,15 +48,46 @@ class ModelCompositor {
...
+          keepOption = pair.sample;
+          keepOption.isKeep = true;
+        } else {
+          let s = suggestionDistribMap[displayText];
```
Comment (eddieantonio) [inline common/predictive-text/worker/model-compositor.ts]: What is `s`? Could you please rename this? An individual suggestion? I'm a bit unclear as to what this branch does.

File: common/predictive-text/worker/models/common.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
+namespace models {
+  export class Common {
```
Comment (eddieantonio) [inline common/predictive-text/worker/models/common.ts]: Does this work if you define it as a `namespace` instead of a class? It isn't really a class...

File: common/predictive-text/worker/models/common.ts [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+    static applyTransform(transform: Transform, context: Context): Context {
```
Comment (eddieantonio) [inline common/predictive-text/worker/models/common.ts]: Haha, I think this helper may have long been overdue!

File: common/predictive-text/worker/models/common.ts [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+      // First, get the current context
+      let fullLeftContext = context.left || '';
+      let lLen = fullLeftContext.length;
+      let lDel = lLen < transform.deleteLeft ? lLen : transform.deleteLeft;
+
+      let leftContext = fullLeftContext.substr(0, lLen - lDel) + (transform.insert || '');
```
Comment (eddieantonio) [inline common/predictive-text/worker/models/common.ts]:
```suggestion
      let leftContext = fullLeftContext.substring(0, lLen - lDel) + (transform.insert || '');
```

`.substr()` is a [legacy function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substr), and its use is discouraged. That's why VSCode autosuggests it when you start typing "myString.subs" ;)

File: common/predictive-text/worker/models/common.ts [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+
+      let fullRightContext = context.right || '';
+      let rLen = fullRightContext.length;
+      let rDel = rLen < transform.deleteRight ? rLen : transform.deleteRight;
+
+      let rightContext = fullRightContext.substr(rDel);
```
Comment (eddieantonio) [inline common/predictive-text/worker/models/common.ts]:
```suggestion
      let rightContext = fullRightContext.substring(rDel);
```

Same reasoning as before.

File: common/predictive-text/worker/models/trie-model.ts [hunk 14]
```diff
@@ -101,13 +102,15 @@
...
+      // Compute the results of the keystroke:
+      let newContext = Common.applyTransform(transform, context);
+
+      // Computes the different in word length after applying the transform above.
+      let leftDelOffset = transform.deleteLeft - transform.insert.length;
+
       // All text to the left of the cursor INCLUDING anything that has
       // just been typed.
-      let prefix = leftContext + (transform.insert || '');
+      let prefix = this.getLastWord(newContext.left);
```
Comment (eddieantonio) [inline common/predictive-text/worker/models/trie-model.ts]: :+1:

File: common/predictive-text/worker/worker-interfaces.ts [hunk 15]
```diff
@@ -38,8 +38,8 @@ type ImportScripts = typeof DedicatedWorkerGlobalScope.prototype.importScripts;
 /**
  * The valid incoming message kinds.
  */
-type IncomingMessageKind = 'config' | 'load' | 'predict' | 'unload';
-type IncomingMessage = ConfigMessage | LoadMessage | PredictMessage | UnloadMessage;
+type IncomingMessageKind = 'config' | 'load' | 'predict' | 'unload' | 'wordbreak';
+type IncomingMessage = ConfigMessage | LoadMessage | PredictMessage | UnloadMessage | WordbreakMessage;
```
Comment (eddieantonio) [inline common/predictive-text/worker/worker-interfaces.ts]: 👍

File: common/predictive-text/message.d.ts [hunk 16]
```diff
@@ -72,6 +72,25 @@ interface SuggestionMessage {
...
+
+  /**
+   * Opaque, unique token that pairs this message
+   * with the wordbreak message that initiated it.
+   */
+  token: Token;
+
+  /**
+   * Ordered array of suggestions, most probable first, least
```
Comment (eddieantonio) [inline common/predictive-text/message.d.ts]: Please update this comment.

File: common/predictive-text/message.d.ts [hunk 17]
```diff
@@ -72,6 +72,25 @@ interface SuggestionMessage {
   suggestions: Suggestion[];
 }
 
+/**
+ * Sends the keyboard an ordered list of suggestions.
```
Comment (eddieantonio) [inline common/predictive-text/message.d.ts]: Please update this comment.

File: common/predictive-text/message.d.ts [hunk 18]
```diff
@@ -257,4 +276,10 @@ interface Suggestion {
    * When suggesting a word, `displayAs` should be that entire word.
    */
   displayAs: string;
+
+  /**
+   * Indicates if this option represents preservation of the user's 
+   * originally-input text (pre-prediction).
+   */
+  isKeep?: boolean;
```
Comment (eddieantonio) [inline common/predictive-text/message.d.ts]: Shall we implement the generalized suggestion tag here? e.g.,

```suggestion
  tag?: 'keep';
```

This is to accommodate more suggestion types such as `emoji` and `correction`. Recall [this document](https://docs.google.com/document/d/18_cJA5_rR6ev1aRUteJC7HrHk4LWOTM1oRk8in3fBKY/edit#heading=h.f91dsrwpvozd). I also think clearer names than `keep` may be `verbatim` or `original`

File: common/predictive-text/message.d.ts (hunk 1 above)
Comment (jahorton) [reply common/predictive-text/message.d.ts]: Sounds reasonable.  I debated the names internally quite a bit, with nothing feeling quite right.  Most of the names I wanted to choose were going to involve reusing the same message name for both directions, which definitely didn't feel right given the rest of our existing design.

File: common/predictive-text/unit_tests/headless/default-word-breaker.js (hunk 3 above)
Comment (jahorton) [reply common/predictive-text/unit_tests/headless/default-word-breaker.js]: It's something quite natural to do when a user places the caret within a word, so I figured that'd be a usability thing we'd want to see happen eventually.

File: common/predictive-text/worker/model-compositor.ts (hunk 7 above)
Comment (jahorton) [reply common/predictive-text/worker/model-compositor.ts]: I'd say to other, non-ASCII wordbreak characters.  Not sure on the best way to handle that at the moment, personally, though that may be since I'm not the one who implemented the default word breaker.

File: common/predictive-text/worker/models/common.ts (hunk 10 above)
Comment (jahorton) [reply common/predictive-text/worker/models/common.ts]: I could leave it as a stray function; that is possible in TypeScript.  I'm probably too used to Java, with its "thou shalt put everything in a class."

File: common/predictive-text/message.d.ts (hunk 18 above)
Comment (jahorton) [reply common/predictive-text/message.d.ts]: Hmm.  That's a pretty reasonable idea, though I'd want to be sure we've thought everything through clearly regarding tags first.  Admittedly, I forgot about the specific document since it's seemed like major back-burner stuff for a future version, so thanks for bringing it up.

* Can a Suggestion have more than one tag?
  - Suppose the text `smile`.
    - `smile` itself is the obvious default 'keep' option.
    - However, 😄 would be a pretty reasonable suggestion, and semantically might be considered a 'keep.'  But yeah, that's probably a 'prediction'.
  - Now consider the text `smilf`.
    - `smile` would be an obvious 'correction'.
    - But what about 😄 here?  It's _both_ an 'emoji' and a 'correction' - do I choose 'emoji' and drop the 'correction' metadata component?
    - If we block 'correction' but allow emojis, we risk the situation where `smilf` does not suggest `smile` (the 'correction') but does suggest 😄 (the 'emoji')... that doesn't seem right.
* I'd prefer to implement the typing here once, so the above point is to ask this - should it be a single string, or would we want an array of tags?

**Comment by jahorton** [comment]
> Additionally, are we still maintaining this document: https://github.com/keymanapp/keyman/blob/master/common/predictive-text/docs/worker-communication-protocol.md?
> 
> If so, this pull request should also add the `wordbreak` and `word` message to that document. If we're not maintaining that document anymore, then we should delete it altogether.

👍 

Totally forgot about that before; thanks for the reminder.

File: common/predictive-text/unit_tests/headless/default-word-breaker.js (hunk 3 above)
Comment (jahorton) [reply common/predictive-text/unit_tests/headless/default-word-breaker.js]: Also, I figured that's why we have LMLayer requesting text to the right-hand side of the caret for part of the `Context` spec.

File: common/predictive-text/message.d.ts (hunk 18 above)
Comment (eddieantonio) [reply common/predictive-text/message.d.ts]: A single string. This will make implementation, for us and for developers of custom models, easier. In this specific example, I think the loss of information— that 😄 is a correction of "smilr"— is okay. I assume that emoji would always be the rightmost suggestion, regardless if it is a "correction" or not. I think the "keep" tag should only be emitted when every suggestion the model(s) make suggest something other than the text that should be kept.

File: common/predictive-text/message.d.ts [hunk 19]
```diff
@@ -41,8 +41,8 @@ type Token = number;
 /**
  * The valid outgoing message kinds.
  */
-type OutgoingMessageKind = 'ready' | 'suggestions';
-type OutgoingMessage = ReadyMessage | SuggestionMessage;
+type OutgoingMessageKind = 'ready' | 'suggestions' | 'currentword';
+type OutgoingMessage = ReadyMessage | SuggestionMessage | CurrentWordMessage;
```
Comment (eddieantonio) [inline common/predictive-text/message.d.ts]: 👍

File: common/predictive-text/unit_tests/headless/default-word-breaker.js [hunk 20]
```diff
@@ -19,4 +20,107 @@ describe('The default word breaker', function () {
...
+
+    // Wordbreaking on a empty space => no word.
+    context = { 
+      left: '', startOfBuffer: true,
+      right: '', endOfBuffer: true
+    };
+
+    broken = model.wordbreak(context);
+
+    assert.strictEqual(broken, '');
```
Comment (eddieantonio) [inline common/predictive-text/unit_tests/headless/default-word-breaker.js]: 👍

File: common/predictive-text/worker/model-compositor.ts (hunk 7 above)
Comment (eddieantonio) [reply common/predictive-text/worker/model-compositor.ts]: Maybe I don't understand the intent of this code. It's checking for a space or newline because it assumes that...? 

If you're looking specifically for whitespace, then use all characters with Unicode general property Z*:

```javascript
/[\u0020\u00a0\u1680\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200a\u2028\u2029\u202f\u205f\u3000]/
```

These DO NOT include CR, LF, and Tab. So you could use this regex to include them:

```javascript
/[\u0009\u000A\u000D\u0020\u00a0\u1680\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200a\u2028\u2029\u202f\u205f\u3000]/
```

We can address this in a future PR, though! 😅

File: common/predictive-text/message.d.ts (hunk 18 above)
Comment (jahorton) [reply common/predictive-text/message.d.ts]: > I think the "keep" tag should only be emitted when every suggestion the model(s) make suggest something other than the text that should be kept.

I disagree with this part, simply to ensure we can reliably detect keep and preserve it as the first option.  Otherwise, there's a slim chance 'keep' could be an obscure and unlikely word and become buried past the first three suggestions.

Otherwise, 👍 .

**Comment by jahorton** [comment]
> LGTM!
> 
> I think we might want to address my remaining comments in future PRs:
> 
>     1. using a generalized "tag" mechanism for suggestions and nailing down _exactly_ what those semantics are
> 
>     2. Spacing in the model compositor

I went ahead and did something basic for them now, though it's quite possible we'd want to address those points further at a later point.

Re "spacing in the model compositor" and the following related concern:

> Maybe I don't understand the intent of this code. It's checking for a space or newline because it assumes that...?

If a user didn't (intentionally) input a whitespace keystroke, we shouldn't give them predictions for entirely new words - predictions that _would_ result if we allow a 'possible alternate' whitespace transform to be the root of possible 'corrections'.  It didn't "feel right" during my development testing after fixing the old backspace-prediction issue to allow them.  Of course, they _should_ be allowed when they're actually intended - so if our 'most likely' input transform is whitespace, we then don't block them.  

**Example case**: With "total" + "l", you'd see the prediction "totally"... and then "the", "and", "of" because there were no other predictions rooted on "total" that didn't start new words.  The intuitive interpretation would likely be "why would I want to replace 'totall' with 'and'?" - something that'd be technically wrong and result in a bad impression for the user.

Similarly, predictions get a little weird if you allow 'correcting' non-backspace keystrokes to a possible backspace keystroke.  It actually does improve the range of correction and prediction, but there was a little bug that was happening for those BKSP-rooted suggestions and the behavior of suggestions didn't quite feel consistent, so leaving it on didn't quite "feel right" either.

If you want to experience the difference, go to /web/testing/prediction-ui/index.html in Chrome (for emulated mobile testing) after disabling the related filtering checks.

File: common/predictive-text/worker/model-compositor.ts [hunk 21]
```diff
@@ -6,15 +6,53 @@ class ModelCompositor {
     this.lexicalModel = lexicalModel;
   }
 
+  protected isWhitespace(transform: Transform): boolean {
+    // Matches prefixed text + any instance of a character with Unicode general property Z* or the following: CR, LF, and Tab.
+    let whitespaceRemover = /.*[\u0009\u000A\u000D\u0020\u00a0\u1680\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200a\u2028\u2029\u202f\u205f\u3000]/iu;
```
Comment (jahorton) [inline common/predictive-text/worker/model-compositor.ts]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/unicode

IE not supported.  :(
